### PR TITLE
Add Model Analyzer to Triton SDK Container

### PR DESCRIPTION
**Pull request by Tabrizian**
It should be merged after 20.11 code freeze.

### CODE REVIEW

File: Dockerfile.sdk [hunk 1]
```diff
@@ -29,12 +29,16 @@ ARG BASE_IMAGE=nvidia/cuda:10.1-devel-ubuntu18.04
 
 ARG TRITON_COMMON_REPO_TAG=main
 ARG TRITON_CORE_REPO_TAG=main
+ARG TRITON_MODEL_ANALYZER_REPO_TAG=main-dev
```
Comment (dzier) [inline Dockerfile.sdk:32]: This will need to be `main` branch before we merge this change

File: Dockerfile.sdk [hunk 2]
```diff
@@ -29,12 +29,16 @@ ARG BASE_IMAGE=nvidia/cuda:10.1-devel-ubuntu18.04
...
 
 FROM ${BASE_IMAGE}
 
 # Ensure apt-get won't prompt for selecting options
 ENV DEBIAN_FRONTEND=noninteractive
 
+# DCGM version to install for Model Analyzer
+ENV DCGM_VERSION=2.0.13
```
Comment (dzier) [inline Dockerfile.sdk:40]: We may have to hold off on this for now. If we go with NVML, then all of this new installation is not needed.

File: Dockerfile.sdk [hunk 3]
```diff
@@ -108,6 +112,19 @@ RUN pip3 install --upgrade numpy pillow && \
     "tritonclient-*-manylinux1_x86_64.whl" | xargs printf -- '%s[all]' | \
     xargs pip3 install --upgrade
 
+# Install DCGM
+RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/datacenter-gpu-manager_${DCGM_VERSION}_amd64.deb && \
+    dpkg -i datacenter-gpu-manager_${DCGM_VERSION}_amd64.deb
+
+# Install Model Analyzer
+ARG TRITON_MODEL_ANALYZER_REPO_TAG
+RUN git clone --single-branch --depth=1 -b $TRITON_MODEL_ANALYZER_REPO_TAG \
```
Comment (dzier) [inline Dockerfile.sdk:121]: If there is a top level setup.ini file, you don't need to clone the repo. You can just use:
```
ARG TRITON_MODEL_ANALYZER_REPO=https://github.com/triton-inference-server/model_analyzer@${TRITON_MODEL_ANALYZER_REPO_TAG}

pip install "git+${TRITON_MODEL_ANALYZER_REPO}"
```

This is one way we will instruct people install the model analyzer.

File: Dockerfile.sdk [hunk 4]
```diff
@@ -108,6 +112,19 @@ RUN pip3 install --upgrade numpy pillow && \
...
+    https://github.com/triton-inference-server/model_analyzer.git && \
+    (export VERSION=`cat /workspace/VERSION` && cd model_analyzer && python3 setup.py install)
 
-ENV PATH //workspace/install/bin:${PATH}
+# Remove Model Analyzer Git source
+RUN rm -rf model_analyzer
+
+ENV PATH /workspace/install/bin:${PATH}
```
Comment (dzier) [inline Dockerfile.sdk:128]: With pip install, this is not needed

**Review by dzier — CHANGES_REQUESTED** [review]
Note that this will be used for how we release Model Analyzer.

Also, this does not rename the Triton client containers, so the PR title should change. You should separate adding Model Analyzer from renaming the container in two different PRs / commits.

File: Dockerfile.sdk (hunk 4 above)
Comment (Tabrizian) [reply Dockerfile.sdk:128]: This was in the Dockerfile before. I think this is because of the `perf_analyzer`.

File: Dockerfile.sdk (hunk 1 above)
Comment (CoderHam) [reply Dockerfile.sdk]: Please add TRITON_MODEL_ANALYZER_REPO_TAG to the gitlab yml (gitlab PR)

Comment (Tabrizian) [reply Dockerfile.sdk]: It is in the gitlab PR.

File: Dockerfile.client [hunk 5]
```diff
@@ -29,12 +29,16 @@ ARG BASE_IMAGE=nvidia/cuda:10.1-devel-ubuntu18.04
...
 ARG TRITON_CORE_REPO_TAG=main
+ARG TRITON_MODEL_ANALYZER_REPO_TAG=main
 
 FROM ${BASE_IMAGE}
 
 # Ensure apt-get won't prompt for selecting options
 ENV DEBIAN_FRONTEND=noninteractive
 
+# DCGM version to install for Model Analyzer
+ENV DCGM_VERSION=2.0.13
```
Comment (deadeyegoodwin) [inline Dockerfile.client:40]: Why not an ARG? Is this envvar needed when the container is running?

File: Dockerfile.client [hunk 6]
```diff
@@ -108,6 +112,15 @@ RUN pip3 install --upgrade numpy pillow && \
     "tritonclient-*-manylinux1_x86_64.whl" | xargs printf -- '%s[all]' | \
     xargs pip3 install --upgrade
 
+# Install DCGM
+RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/datacenter-gpu-manager_${DCGM_VERSION}_amd64.deb && \
+    dpkg -i datacenter-gpu-manager_${DCGM_VERSION}_amd64.deb
+
+# Install Model Analyzer
+ARG TRITON_MODEL_ANALYZER_REPO_TAG
+ENV TRITON_MODEL_ANALYZER_REPO "https://github.com/triton-inference-server/model_analyzer@${TRITON_MODEL_ANALYZER_REPO_TAG}"
```
Comment (deadeyegoodwin) [inline Dockerfile.client:121]: Why not an ARG? Is this envvar needed when the container is running?

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Review by deadeyegoodwin — APPROVED** [review]

**Pull request merged**